services/processor/tasks/genesis/genesis.go

```diff
@@ -0,0 +1,211 @@
+	pool := work.NewWorkerPool(ProcessGenesisSingletonTask{}, concurrency, poolName, redisPool)
+	queue := work.NewEnqueuer(poolName, redisPool)
+
+	// https://github.com/gocraft/work/issues/10#issuecomment-237580604
+	// adding fields via a closure gives the workers access to the lotus api, a global could also be used here
+	pool.Middleware(func(mt *ProcessGenesisSingletonTask, job *work.Job, next work.NextMiddlewareFunc) error {
+		mt.node = node
+		mt.pubCh = pubCh
+		mt.log = logging.Logger("genesistask")
+		return next()
+	})
+	logging.SetLogLevel("genesistask", "info")
+	// log all task
+	pool.Middleware((*ProcessGenesisSingletonTask).Log)
+
+	// register task method and don't allow retying
+	pool.JobWithOptions(taskName, work.JobOptions{
+		MaxFails: 1,
+	}, (*ProcessGenesisSingletonTask).Task)
+
+	return pool, queue
+}
+
+type ProcessGenesisSingletonTask struct {
+	node lapi.FullNode
+	log  *logging.ZapEventLogger
+
+	pubCh chan<- model.Persistable
+
+	genesis   types.TipSetKey
+	stateroot cid.Cid
+}
+
+func (p *ProcessGenesisSingletonTask) Log(job *work.Job, next work.NextMiddlewareFunc) error {
+	p.log.Infow("Starting Job", "name", job.Name, "Args", job.Args)
+	return next()
+}
+
+func (p *ProcessGenesisSingletonTask) ParseArgs(job *work.Job) error {
+	srStr := job.ArgString("stateroot")
+	if err := job.ArgError(); err != nil {
+		return err
+	}
+
+	tsStr := job.ArgString("ts")
+	if err := job.ArgError(); err != nil {
+		return err
+	}
+
+	sr, err := cid.Decode(srStr)
+	if err != nil {
+		return err
+	}
+
+	var tsKey types.TipSetKey
+	if err := tsKey.UnmarshalJSON([]byte(tsStr)); err != nil {
+		return err
+	}
+	p.genesis = tsKey
+	p.stateroot = sr
+	return nil
+}
+
+func (p *ProcessGenesisSingletonTask) Task(job *work.Job) error {
+	if err := p.ParseArgs(job); err != nil {
+		return err
+	}
+	ctx := context.TODO()
+
+	genesisAddrs, err := p.node.StateListActors(ctx, p.genesis)
+	if err != nil {
+		return err
+	}
+
+	result := &genesismodel.ProcessGenesisSingletonResult{}
+	for _, addr := range genesisAddrs {
+		genesisAct, err := p.node.StateGetActor(ctx, addr, p.genesis)
+		if err != nil {
+			return err
+		}
+		switch genesisAct.Code {
+		case builtin.SystemActorCodeID:
+			// TODO
+		case builtin.InitActorCodeID:
+			// TODO
+		case builtin.CronActorCodeID:
+			// TODO
+		case builtin.AccountActorCodeID:
+			// TODO
+		case builtin.StoragePowerActorCodeID:
+			// TODO
+		case builtin.StorageMarketActorCodeID:
+		case builtin.StorageMinerActorCodeID:
+			res, err := p.storageMinerState(ctx, addr, genesisAct)
+			if err != nil {
+				return err
+			}
+			result.AddMiner(res)
+		case builtin.PaymentChannelActorCodeID:
+			// TODO
+		case builtin.MultisigActorCodeID:
+			// TODO
+		case builtin.RewardActorCodeID:
+			// TODO
+		case builtin.VerifiedRegistryActorCodeID:
+			// TODO
+		default:
+			return fmt.Errorf("unknown actor in genesis state. address: %s code: %s head: %s", addr, genesisAct.Code, genesisAct.Head)
```

Comment:
Seems drastic to abort just because we have an unknown actor. Maybe log and continue.